### PR TITLE
Overhaul attachment downloading sagas + state

### PR DESCRIPTION
**Pull request by chromakode**
These changes fix fighting progress bars on download. They also fix cases where the same file (hdPreview/download) would be downloaded multiple times concurrently.

 * Unify hdPreview and download cache / progress.
 * Remove 'downloading-preview', 'downloading' `AttachmentMessage` states.
 * Handle load actions when already loading gracefully.
 * Use separate progress states for previews and download.

Tested on desktop, iOS, and Android.

### CODE REVIEW

File: shared/reducers/chat.js [hunk 1]
```diff
@@ -250,16 +277,7 @@ function reducer (state: Constants.State = initialState, action: Constants.Actio
...
-      // $FlowIssue
-      return state.update('conversationStates', conversationStates => updateConversationMessage(
-        conversationStates,
-        conversationIDKey,
-        item => !!item.messageID && item.messageID === messageID,
-        m => ({
-          ...m,
-          ...message,
-        })
-      )).set('messageMap', state.get('messageMap').set(message.key, message))
```
Comment (chromakode) [inline shared/reducers/chat.js]: This was buggy -- it doesn't merge the message properties in the messageMap. Using `updateStateWithMessageChanged` fixes this.

File: shared/reducers/chat.js [hunk 2]
```diff
@@ -104,6 +104,33 @@ function updateConversationMessage (conversationStates: ConversationsStates, con
   )
 }
 
+function updateStateWithMessageChanged (state: Constants.State, conversationIDKey: Constants.ConversationIDKey, messageID: Constants.MessageID, toMerge: $Shape<Constants.Message>) {
```
Comment (chromakode) [inline shared/reducers/chat.js]: Useful for now, hopefully can be thrown out when we refactor w/ entities.

File: shared/package.json [hunk 3]
```diff
@@ -67,6 +67,7 @@
     "menubar": "5.1.0",
     "mkdirp": "0.5.1",
     "moment": "2.18.0",
+    "path-parse": "^1.0.5",
```
Comment (chrisnojima) [inline shared/package.json]: remove ^

**Comment by chrisnojima** [comment]
i need this for a pr i'm making so i'm merging

**Pull request merged**